src/firejail/template.c

```diff
@@ -0,0 +1,429 @@
+
+		switch (type) {
+		case STR_CHECK_ALNUM:
+			if (!isalnum(str[i]))
+				return -EINVAL;
+
+			break;
+		case STR_CHECK_DBUS:
+			/* Allow '_./' in str */
+			if (!isalnum(str[i]) && str[i] != '_' &&
+					str[i] != '.' && str[i] != '/')
+				return -EINVAL;
+
+			/* Do not allow '..', prevents also '../'.*/
+			if (str[i] == '.' && str[i+1] == '.')
+				return -EINVAL;
+
+			break;
+		}
+	}
+
+	return 0;
+}
+
+void check_template(const char *arg) {
+	Template *tmpl;
+	const char *key;
+	const char *value;
+	char *delim = TEMPLATE_KEY_VALUE_DELIM;
+	char *macro = NULL;
+	char *macro_value = NULL;
+	int err = 0;
+
+	fprintf(stdout, "Check %s list %p\n", arg, tmpl_list);
+
+	/* Only alphanumeric chars in template key. */
+	key = strtok((char*)arg, delim);
+	if (is_valid_str(key, STR_CHECK_ALNUM) == -EINVAL) {
+		fprintf(stderr, "Error invalid template key \"%s\"", key);
+		exit(1);
+	}
+
+	/* Only a-zA-Z0-9_ */
+	value = strtok(NULL, delim);
+	if (is_valid_str(value, STR_CHECK_DBUS) == -EINVAL) {
+		fprintf(stderr, "Error invalid template value in \"%s:%s\"",
+								key, value);
+		exit(1);
+	}
+
+	if (asprintf(&macro, "${%s}", key) == -1)
+		errExit("asprintf");
+
+	/* Hardcoded macro or XDG value is not allowed to be overridden. */
+	macro_value = resolve_macro(macro);
+	if (macro_value) {
+		fprintf(stderr, "Error override of \"${%s}\" is not permitted",
+									key);
+		err = -EPERM;
+		free(macro_value);
+		goto out;
+	}
+
+	tmpl = template_new(key, value);
+	err = template_add(tmpl);
+
+	switch (err) {
+	case 0:
+		goto out;
+	case -EINVAL:
+		fprintf(stderr, "Error invalid template key \"%s\" "
+						"value \"%s\"\n", key, value);
+		break;
+	case -EALREADY:
+		fprintf(stderr, "Error template key \"%s\" already exists\n",
+								key);
+		break;
+	}
+
+	template_free(tmpl);
+
+out:
+	free(macro);
+
+	if (err)
+		exit(-err);
+}
+
+static char* append_to_string(char *str1, const char *str2)
+{
+	size_t len;
+
+	if (!str2)
+		return str1;
+
+	if (!str1)
+		return strdup(str2);
+
+	len = strlen(str2) + 1;
+	str1 = (char*)realloc(str1, strlen(str1) + len);
+
+	return strncat(str1, str2, len);
+}
+
+const char *internal_keys[] = { "HOME", "CFG", "RUNUSER", "PATH", "PRIVILEGED",
+				NULL };
+
+static int is_internal_macro(const char *key)
+{
+	char *macro;
+	char *macro_value;
+	int err = 0;
+	int i;
+
+	for (i = 0; internal_keys[i]; i++) {
+		if (!strncmp(key, internal_keys[i], strlen(internal_keys[i]))) {
+			err = -EEXIST;
+			goto exit;
+		}
+	}
+
+	if (asprintf(&macro, "${%s}", key) == -1)
+		errExit("asprintf");
+
+	macro_value = resolve_macro(key);
+	if (macro_value) {
+		err = -EEXIST;
+		free(macro_value);
+	}
+
+	free(macro);
+exit:
+	return err;
+}
+
+static char *process_key_value(char *container, char *key, int *err)
+{
+	char *str;
+	char *token;
+	char *saveptr;
+	const char *delim = TEMPLATE_KEY_MACRO_SUB_DELIMS;
+	const char *value;
+
+	for (str = key; ; str = NULL) {
+		token = strtok_r(str, delim, &saveptr);
```

Comment:
This splits at '{' or '}' - whichever comes first. Which would tokenize "foo  $bar}baz{" - how exactly?

Reply:
That is a bug that can be easily fixed.